File: BirthdayBot/BackgroundServices/GuildStatistics.vb
Imports System.Net

Class GuildStatistics
    Inherits BackgroundService

    Private ReadOnly Property DBotsToken As String

    Public Sub New(instance As BirthdayBot)
        MyBase.New(instance)
        DBotsToken = instance.Config.DBotsToken
    End Sub

    Public Overrides Async Function OnTick(tick As Integer) As Task
        ' Activate roughly every 2 hours (interval: 45)
        If tick Mod 160 <> 2 Then Return

        Dim count = BotInstance.DiscordClient.Guilds.Count
        Log($"Currently in {count} guild(s).")

        Await SendExternalStatistics(count)
    End Function

    ''' <summary>
    ''' Send statistical information to external services.
    ''' </summary>
    ''' <remarks>
    ''' Only Discord Bots is currently supported. No plans to support others any time soon.
    ''' </remarks>
    Async Function SendExternalStatistics(guildCount As Integer) As Task
        Dim rptToken = BotInstance.Config.DBotsToken
        If rptToken Is Nothing Then Return

        Const apiUrl As String = "https://discord.bots.gg/api/v1/bots/{0}/stats"
        Using client As New WebClient()
            Dim uri = New Uri(String.Format(apiUrl, CType(BotInstance.DiscordClient.CurrentUser.Id, String)))
            Dim data = "{ ""guildCount"": " + CType(guildCount, String) + " }"
            client.Headers(HttpRequestHeader.Authorization) = rptToken
            client.Headers(HttpRequestHeader.ContentType) = "application/json"
            Try
                Await client.UploadStringTaskAsync(uri, data)
                Log("Discord Bots: Report sent successfully.")
            Catch ex As WebException
                Log("Discord Bots: Encountered an error. " + ex.Message)
            End Try
        End Using
    End Function
End Class

File: BackgroundServices/GuildStatistics.cs
using System;
using System.Net.Http;
using System.Text;
using System.Threading.Tasks;

namespace BirthdayBot.BackgroundServices
{
    class GuildStatistics : BackgroundService
    {
        private static readonly HttpClient _httpClient = new HttpClient();

        public GuildStatistics(BirthdayBot instance) : base(instance) { }

        public async override Task OnTick()
        {
            var count = BotInstance.DiscordClient.Guilds.Count;
            Log($"Currently in {count} guilds.");
            await SendExternalStatistics(count);
        }

        /// <summary>
        /// Send statistical information to external services.
        /// </summary>
        async Task SendExternalStatistics(int count)
        {
            var dbotsToken = BotInstance.Config.DBotsToken;
            if (dbotsToken != null)
            {
                const string dBotsApiUrl = "https://discord.bots.gg/api/v1/bots/{0}/stats";
                const string Body = "{{ \"guildCount\": {0} }}";
                var uri = new Uri(string.Format(dBotsApiUrl, BotInstance.DiscordClient.CurrentUser.Id));

                var post = new HttpRequestMessage(HttpMethod.Post, uri);
                post.Headers.Add("Authorization", dbotsToken);
                post.Content = new StringContent(string.Format(Body, count), Encoding.UTF8, "application/json");

                await Task.Delay(80); // Discord Bots rate limit for this endpoint is 20 per second
                await _httpClient.SendAsync(post);
                Log("Discord Bots: Count sent successfully.");
            }
        }
    }
}
